First version

Krulknul opened the pull request:
First version of the client after a refactor to use a builder pattern for building requests.
There are a few, but insufficient tests at present. If this gets merged, I can write more tests.

Krulknul committed:
initial push

Krulknul committed:
make core and gateway optional features each

Krulknul committed:
enable all features by default in vs-code

Krulknul committed:
make clippy just a bit happier

Krulknul committed:
remove vs-code feature setting, as the features are already default in cargo.toml

Krulknul committed:
Add Debug trait to stream_type struct

Krulknul committed:
add chrono for timestamps

Krulknul committed:
remove unwrap

Krulknul committed:
remove unwrp

Krulknul committed:
remove another unwrap

Krulknul committed:
improve error handling and move to builder pattern for gateway

Krulknul committed:
give core a builder pattern

Krulknul committed:
fix import issues

Krulknul committed:
change request naming

Krulknul committed:
update readme

Krulknul committed:
update readme

Krulknul committed:
Refactor README.md to include blocking and async examples for using the GatewayClient

Krulknul committed:
add comment

Krulknul committed:
update gitignore

Krulknul committed:
update readme

Krulknul committed:
remove empty tests directory

fpieper commented on README.md:
```diff
@@ -1,2 +1,60 @@
+
+// Use a builder pattern to create and execute the request
+let response = client
+    .get_transactions_stream_builder()
```
Normally by convention `get_` is not common in Rust.

Krulknul replied:
let's remove that, but keep the builder suffix

fpieper commented on README.md:
```diff
@@ -1,2 +1,60 @@
+let response = client
+    .get_transactions_stream_builder()
+    .order(Order::Asc)
+    .limit_per_page(10)
```
Does the builder pattern work well with querying multiple pages using the cursor?

Do we have an alternative to `execute`? Maybe `fetch` or something you would always need to decide like `json()` or `xml()` (I now we don't have xml ;) )?

fpieper commented on src/lib.rs:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    #[maybe_async_attr]
+    pub async fn post_request<S: Serialize>(
```
I would probably name it `post`, but is mainly internal anyway 👍

fpieper commented on src/lib.rs:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    #[maybe_async_attr]
+    pub async fn post_request<S: Serialize>(
```
Here too `post`

fpieper commented on src/gateway/stream/get_transactions_stream.rs:
```diff
@@ -0,0 +1,388 @@
+    }
+
+    pub fn with_receipt_state_changes(&mut self) -> &mut Self {
+        self.request.opt_ins = match self.request.opt_ins.clone() {
```
For the Opt-Ins there seems to be a lot of boilerplate? I guess the code can be reduced? Using `map`, and/or default on `Option`

Krulknul replied:
yes, will change this to be shorter

fpieper reviewed (commented):
Thank you very much. Don't have lots of feedback since it is looking good overall 👍 and the structure is mostly defined by the underlaying API

Krulknul committed:
changes as discussed:
remove get prefix from endpoints
change post_request function name
chante execute to fetch
remove opt-ins code duplication

Krulknul committed:
bump version to 1.0.0
